### PR TITLE
[kml] Fix bookmark names lost on GeoJSON and GPX export

### LINKED ISSUE #13237: Export gps tags with Russian comments
existing issues?

Describe the issue
When I add a bookmark with a GPS point and manually enter a custom name (with a description/comment), then export it to GeoJSON, the exported name field shows the date the point was added instead of the custom name I typed.

Steps to reproduce

1. Place a point (bookmark) on the map.
2. Enter a custom name with my description/comment in the name field.
3. Save the bookmark to one of my catalogs.
4. Export the bookmark to GeoJSON.
5. Open the exported GeoJSON file. The name field for that point contains only the date it was added, not the custom name I entered.

Expected behaviour
The exported GeoJSON file should retain the manually entered name in the name field, as it did before today (previously the custom name was preserved correctly).

Screenshots
N/A

System information:

· Operating system and its version: iOS (latest version)
· Organic Maps version: latest version
· Device Model: iPhone 16 Pro Max

Additional context
This appears to be a recent regression. Until now, exporting to GeoJSON always included the custom name I had typed. Now it only outputs the addition date instead of that name.

### PR DESCRIPTION
**Pull request by biodranik**
Fixes #13237.

A bookmark stores the name it was created with in `m_name` and the user-entered name in `m_customName`. GeoJSON exported only `m_name`, so renamed bookmarks—and bookmarks created at the current position—could be exported under an old name or creation date.

## What changed

- KML, GPX and GeoJSON bookmark writers now use `GetPreferredBookmarkName(bookmark, "default")`: custom name first, then original name, then the localized feature type.
- The `"default"` resolver path selects localized strings deterministically as `default` → `int_name` → `en` → lowest remaining language code. `alt_name` and `old_name` rank behind every real language.
- The resolver keeps comma-separated `int_name` values intact for serialization instead of applying feature-rendering truncation.
- Bookmark/category/track names and descriptions no longer disappear when they contain no default or matching device language. The UI and region-aware display paths use the same deterministic last resort.
- GeoJSON omits properties with no value. GPX omits an empty bookmark name; KML retains its existing empty `<name></name>` behavior.
- A shared category file uses the resolved category name instead of falling back to its internal on-disk basename.
- The unused `GpxParser::m_customName` state and its unreachable branch were removed.

No platform-specific bridge code changes are needed: the shared C++ implementation is used by iOS, Android and desktop.

## KML round trips

Plain KML `<name>` and `<description>` elements are interoperability projections for readers that do not understand Organic Maps extended data. KML carries no provenance that distinguishes such a projection from an explicit default value, so a round trip retains the selected projection as `default` while preserving all extended translations.

`Kml_RoundTrip_PromotesFallbackToDefaultLanguage` covers this behavior for category, bookmark and track names and descriptions with multiple unmatched languages.

## Validation

- Full arm64 Debug `kml_tests` suite with `CMAKE_UNITY_BUILD=OFF`.
- Unity builds of `kml_tests` and `map_tests`.
- Direct coverage for custom/original/type precedence, complete `int_name` preservation, language and region fallbacks, pseudo-name ranking, KML promotion, and KML/GPX/GeoJSON output.
- `ExportSingleMultilingualCategoryName` exercises the complete `BookmarkManager` sharing path and exported basename.
- `clang-format` and `git diff --check`.

LLM assistance: Codex was used for review, implementation and test design.

### CODE REVIEW

File: libs/kml/serdes.cpp [hunk 1]
```diff
@@ -430,14 +430,13 @@ void SaveBookmarkData(Writer & writer, BookmarkData const & bookmarkData)
 {
   writer << kIndent2 << "<Placemark>\n";
   writer << kIndent4 << "<name>";
-  auto const defaultLang = StringUtf8Multilang::GetLangByCode(kDefaultLangCode);
-  SaveStringWithCDATA(writer, GetPreferredBookmarkName(bookmarkData, defaultLang));
+  SaveStringWithCDATA(writer, GetNameForExport(bookmarkData));
   writer << "</name>\n";
 
-  if (auto const description = GetDefaultLanguage(bookmarkData.m_description))
+  if (auto const description = GetStringForExport(bookmarkData.m_description); !description.empty())
```
Comment (biodranik) [inline libs/kml/serdes.cpp:434]: Writing the plain `<description>`/`<name>` for strings that have no default-language entry promotes them to `default` on the next load: the parser stores those elements under `kDefaultLang` unconditionally (`serdes.cpp:1320` for the placemark description, `:1164`/`:1166` for the Document), and `<ExtendedData>` is written *after* them, so it cannot override.

Measured on this branch vs `master`, saving a `{ru: …}` category/bookmark/track and reloading it:

```
                       master                  this PR
category name          ru=…                    ru=… + default=…
category description   ru=…                    ru=… + default=…
bookmark description   ru=…                    ru=… + default=…
track name             ru=…                    ru=… + default=…
track description      ru=…                    ru=… + default=…
```

(the bookmark `<name>` already behaved this way, so only the other five are new.)

`bookmarks.kml` is OM's own storage, so this runs on every save/load cycle, not only on explicit export — a `name:ru`-only string permanently loses its language tag. Making 3rd-party viewers show *something* is clearly worth it; just worth confirming the promotion is an accepted side effect rather than a surprise.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes_common.cpp [hunk 2]
```diff
@@ -78,12 +79,39 @@ void SaveStringWithCDATA(Writer & writer, std::string const & s)
...
+    return it != lstr.end() && !it->second.empty() ? &it->second : nullptr;
+  };
+
+  for (auto const lang : {kDefaultLang, StringUtf8Multilang::kInternationalCode, StringUtf8Multilang::kEnglishCode})
+    if (auto const value = getNonEmpty(lang))
+      return *value;
+
+  std::string const * fallback = nullptr;
+  int8_t fallbackLang = StringUtf8Multilang::kMaxSupportedLanguages;
+  for (auto const & [lang, value] : lstr)
```
Comment (biodranik) [inline libs/kml/serdes_common.cpp]: Two notes on the last-resort scan.

**1. It can pick `alt_name` (53) or `old_name` (55).** `feature::GetBestName` never can, since neither is in its priority list. A POI tagged `old_name` plus `name:kk`/`name:hi` (codes 56/63) and no plain `name` would be exported under its *old* name. `&& !StringUtf8Multilang::IsAltOrOldName(lang)` in the condition skips them, and `GetNameForExport` then falls through to the localized type, which is more honest.

**2. The policy is a single ordering**, so one pass expresses it in half the lines and drops the lambda + the second loop:

```cpp
std::string GetStringForExport(LocalizableString const & lstr)
{
  auto const priority = [](int8_t lang)
  {
    switch (lang)
    {
    case kDefaultLang: return -3;
    case StringUtf8Multilang::kInternationalCode: return -2;
    case StringUtf8Multilang::kEnglishCode: return -1;
    // Any other language is a deterministic last resort, ordered by its stable code.
    default: return static_cast<int>(lang);
    }
  };

  std::string const * best = nullptr;
  int bestPriority = 0;
  for (auto const & [lang, value] : lstr)
    if (!value.empty() && (best == nullptr || priority(lang) < bestPriority))
    {
      best = &value;
      bestPriority = priority(lang);
    }

  return best ? *best : std::string{};
}
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes_common.hpp [hunk 3]
```diff
@@ -22,7 +22,13 @@ std::string PointToGxString(geometry::PointWithAltitude const & pt);
 bool LineHasAltitude(TrackGeometry const & line);
 
 void SaveStringWithCDATA(Writer & writer, std::string const & s);
-std::string const * GetDefaultLanguage(LocalizableString const & lstr);
+
+// Name/description to write into an exported file, shared by all exporters (KML, GPX, GeoJSON).
+// A name typed by the user is stored in m_customName and must win over the original (POI) name
+// kept in m_name. Strings prefer default/int_name/en; if none exists, the lowest language code is
+// used as a deterministic last resort so that a non-empty localized value is never dropped.
```
Comment (biodranik) [inline libs/kml/serdes_common.hpp]: The comment explains the language ladder but not the part that is actually new for GPX/GeoJSON: a bookmark with no name at all now falls back to `GetLocalizedFeatureType()`, which reads the classificator and the device UI language — so the same collection exports as `Castle` or `Замок` depending on the phone (and on Android costs one JNI resource lookup per nameless bookmark). One sentence here would cover the non-obvious half of `GetNameForExport`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes_gpx.cpp [hunk 4]
```diff
@@ -291,13 +290,10 @@ void GpxParser::Pop(std::string_view tag)
...
-        if (!m_customName.empty())
-          data.m_customName[kDefaultLang] = std::move(m_customName);
-        else if (!data.m_name.empty())
-        {
-          // Here we set custom name from 'name' field for KML-files exported from 3rd-party services.
-          data.m_customName = data.m_name;
-        }
+        // A GPX file carries no OM extended data, so its <name> can only be a user-authored one.
+        // Keep it in m_customName too, as the KML parser does, so that it always wins over a name
+        // OM itself could derive later, see GetPreferredBookmarkName.
```
Comment (biodranik) [inline libs/kml/serdes_gpx.cpp]: "so that it always wins over a name OM itself could derive later" doesn't match the code: nothing re-derives `m_name` for an existing bookmark (rename goes through `Bookmark::SetCustomName`, `libs/map/bookmark.cpp:269`), and here `m_customName == m_name`, so there is nothing to win over.

The accurate rationale is the one the copy actually serves: `m_customName` is what the place page, search and every exporter treat as *the user's* name, so a human-authored GPX `<name>` belongs there — and a later rename replaces it instead of leaving two competing names.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/kml_tests/geojson_tests.cpp [hunk 5]
```diff
@@ -1515,4 +1516,90 @@ UNIT_TEST(GeoJson_Writer_Elevation_AllDefault)
...
+}
+
+UNIT_TEST(GeoJson_Writer_NameFallbacks)
+{
+  classificator::Load();
+
+  // Nameless bookmarks are exported with their localized type, as KML and GPX do.
+  kml::BookmarkData typed;
+  typed.m_featureTypes = {classif().GetTypeByPath({"historic", "castle"})};
+  TEST_EQUAL(ExportedName(typed), kml::GetLocalizedFeatureType(typed.m_featureTypes), ());
```
Comment (biodranik) [inline libs/kml/kml_tests/geojson_tests.cpp]: This compares the writer's output with the very function the writer calls, so it holds for any fallback the writer might pick. In the test binary `platform::GetLocalizedTypeName` is the identity (`libs/platform/localization_dummy.cpp`), so `TEST_EQUAL(ExportedName(typed), "historic-castle", ())` pins the behaviour and documents it at the same time.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by biodranik — COMMENTED** [review]
This is an experimental LLM code review

Reviewed all three commits and every caller that resolves a bookmark name. The core fix is right: `Bookmark::SetCustomName()` is the only thing the rename UI calls and it writes `m_customName` (`libs/map/bookmark.cpp:269`), so exporting `m_name` alone was indeed the #13237 bug, and routing all three writers through one helper removes the divergence. Commit 3 is a genuine no-op: `GpxParser::m_customName` was only ever cleared, and `data.m_customName = data.m_name` on an empty map is equivalent to the old guarded branch. Built and ran `kml_tests` on Linux — 111/111 pass.

Inline comments cover one behaviour change worth confirming (KML now promotes non-default names/descriptions to `default` on reload), one edge case in the fallback scan, a shorter form of `GetStringForExport`, and two comment/test accuracy points. Two more things, both outside the diff:

**Same class of bug, not fixed.** `GetFileNameForExport()` (`libs/map/bookmark_manager.cpp:87`) still uses `kml::GetDefaultStr(m_categoryData.m_name)`, so a category whose name has no default-language entry — exactly the case this PR fixes *inside* the files — is exported as `<source-file-basename>.gpx` instead of under its name. `GetStringForExport()` there would be consistent.

**Placement.** `GetNameForExport()` is the same ladder as `kml::GetPreferredBookmarkName()` (`libs/kml/type_utils.cpp:67`) — custom name → name → localized type — differing only in the string resolver. Both, together with `GetLocalizedFeatureType()` and `GetDefaultStr()`, are name resolution rather than serialization; keeping the two new helpers in `type_utils.{hpp,cpp}` would put the whole policy in one file and let `serdes_common.cpp` drop its new `kml/types.hpp` include.

On alternatives: resolving through `feature::GetBestName()` + `MakeLanguagesPriorityList(kDefaultCode, true)` would reuse the indexer's single source of truth for the priority order, but it needs a `LocalizableString` → `StringUtf8Multilang` conversion per string (that is what makes `GetPreferredBookmarkStr` expensive) and it truncates `int_name` at the first comma, which is wrong for user-authored data. The hand-rolled resolver here is the better trade — worth keeping.

No user-facing strings or search categories changed, so there is nothing to review on the translation side.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes_common.cpp [hunk 6]
```diff
@@ -78,12 +79,39 @@ void SaveStringWithCDATA(Writer & writer, std::string const & s)
     writer << *clean;
 }
 
-std::string const * GetDefaultLanguage(LocalizableString const & lstr)
+std::string GetStringForExport(LocalizableString const & lstr)
```
Comment (biodranik) [inline libs/kml/serdes_common.cpp]: `GetDefaultLanguage` returned a pointer into the map and `SaveStringWithCDATA` streamed it directly; `GetStringForExport` returns by value, so every exported name and description is now heap-copied (2 per bookmark + 2 per track, for the whole collection).

Returning a view and making it a single pass removes the copies and ~10 lines:

```cpp
// serdes_common.hpp
void SaveStringWithCDATA(Writer & writer, std::string_view s);
std::string_view GetStringForExport(LocalizableString const & lstr);

// serdes_common.cpp
std::string_view GetStringForExport(LocalizableString const & lstr)
{
  // Lower rank wins: default, int_name, en, then the lowest language code as a
  // deterministic last resort so a non-empty localized value is never dropped.
  auto const rank = [](int8_t lang) -> int
  {
    if (lang == kDefaultLang)
      return 0;
    if (lang == StringUtf8Multilang::kInternationalCode)
      return 1;
    if (lang == StringUtf8Multilang::kEnglishCode)
      return 2;
    return 3 + lang;
  };

  std::string_view best;
  int bestRank = std::numeric_limits<int>::max();
  for (auto const & [lang, value] : lstr)
    if (!value.empty())
      if (int const r = rank(lang); r < bestRank)
        bestRank = r, best = value;
  return best;
}

std::string GetNameForExport(BookmarkData const & bmData)
{
  if (auto const name = GetStringForExport(bmData.m_customName); !name.empty())
    return std::string{name};
  if (auto const name = GetStringForExport(bmData.m_name); !name.empty())
    return std::string{name};
  return GetLocalizedFeatureType(bmData.m_featureTypes);
}
```

Inside `SaveStringWithCDATA` only `std::string const * clean = &s` → `std::string_view clean = s` changes; the two GeoJSON `description`/`name` assignments need an explicit `std::string{...}`. All call sites keep the `if (auto const x = ...; !x.empty())` shape. Checked locally: builds and all `kml_tests` pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes_common.cpp [hunk 7]
```diff
@@ -78,12 +79,39 @@ void SaveStringWithCDATA(Writer & writer, std::string const & s)
...
+    auto const it = lstr.find(lang);
+    return it != lstr.end() && !it->second.empty() ? &it->second : nullptr;
+  };
+
+  for (auto const lang : {kDefaultLang, StringUtf8Multilang::kInternationalCode, StringUtf8Multilang::kEnglishCode})
+    if (auto const value = getNonEmpty(lang))
+      return *value;
+
+  std::string const * fallback = nullptr;
+  int8_t fallbackLang = StringUtf8Multilang::kMaxSupportedLanguages;
```
Comment (biodranik) [inline libs/kml/serdes_common.cpp]: The lowest-code last resort can pick an OSM pseudo-name over a real one. `Info::FormatNewBookmarkName()` copies *every* language of the feature into `m_name`, and `kLanguages` contains `alt_name` = 53 and `old_name` = 55, below `kk` 56, `mr` 57, `et` 58, `ku` 59, `mn` 60, `mk` 61, `lv` 62, `hi` 63.

So a POI tagged `old_name` + `name:hi` with no plain `name` is exported as the old name. Ranking those two after real languages is enough:

```cpp
if (lang == StringUtf8Multilang::GetLangIndex("alt_name") || lang == StringUtf8Multilang::GetLangIndex("old_name"))
  return 100 + lang;
```

(Also worth knowing: the ordering below `int_name`/`en` is the legacy binary encoding order — `ja` 2 < `fr` 3 < `ko_rm` 4 < `ar` 5 < `de` 6 — so "de wins over ru" in the tests is deterministic but not meaningful to a user.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes_gpx.cpp [hunk 8]
```diff
@@ -291,13 +290,10 @@ void GpxParser::Pop(std::string_view tag)
...
+        data.m_customName = data.m_name;
```
Comment (biodranik) [inline libs/kml/serdes_gpx.cpp:298]: Two notes on this line:

1. The comment says "as the KML parser does", but `serdes.cpp` copies only under four conditions (`m_name.size() == 1 && default lang && m_customName.empty() && m_featureTypes.empty()`). All four hold by construction here (the GPX parser only ever writes `m_name[kDefaultLang]` and never sets feature types), so the behaviour matches — but the comment reads as if the two parsers share a rule they don't.

2. Now that `GetNameForExport()` (and `GetPreferredBookmarkName()`) fall back to `m_name`, is the copy still buying anything? Every consumer resolves `m_customName` → `m_name` → type, so the copy only stores each waypoint name twice in the `.kmb`, and re-exports it twice in KML (`<mwm:name>` + `<mwm:customName>`). Dropping the line entirely looks equivalent — worth double-checking against the "3rd-party services" case the KML rule was written for.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/kml_tests/geojson_tests.cpp [hunk 9]
```diff
@@ -1515,4 +1516,90 @@ UNIT_TEST(GeoJson_Writer_Elevation_AllDefault)
...
+  TEST(!error, (glz::format_error(error, json)));
+  if (error || exported.features.size() != 1)
+    return std::nullopt;
+
+  auto const & properties = exported.features.front().properties;
+  auto const name = properties.find("name");
+  if (name == properties.end())
+    return std::nullopt;
+  TEST(name->second.is_string(), ());
+  return name->second.is_string() ? std::optional<std::string>{name->second.get_string()} : std::nullopt;
```
Comment (biodranik) [inline libs/kml/kml_tests/geojson_tests.cpp]: `base::OnTestFailed` is `[[noreturn]]`, so both re-checks after a `TEST` are unreachable: `error ||` on line 1531 and the ternary on line 1539. They can be:

```cpp
  TEST(!error, (glz::format_error(error, json)));
  if (exported.features.size() != 1)
    return std::nullopt;
  ...
  TEST(name->second.is_string(), ());
  return name->second.get_string();
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by biodranik — COMMENTED** [review]
This is an experimental LLM code review.

The fix itself is correct and well covered: `m_customName` first is exactly what #13237 asks for, and the four rewritten `Marcador de prueba` → `Mi lugar favorito` expectations were indeed pinning the bug (the fixture sets both fields). Verified locally on Linux/GCC 13: `kml_tests` builds clean and all tests pass, including each new test standalone under `--filter`. The three commits are cleanly separated, and commit 3 is a true no-op (`GpxParser::m_customName` was only ever `clear()`ed, and assigning an empty `m_name` to an empty `m_customName` is a no-op).

Inline comments cover: an avoidable copy per exported string, the `alt_name`/`old_name` hole in the last-resort ordering, the GPX parser comment/copy, and dead post-`TEST` checks in the new helper.

Three things that don't fit inline:

**Two `GetDefaultStr` call sites left behind, same bug class as this PR fixes.**
- `serdes_geojson.cpp:470` — the "track without geometry" warning still logs `GetDefaultStr(track.m_name)`, now inconsistent with the exporter two functions below.
- `libs/map/bookmark_manager.cpp:87` `GetFileNameForExport()` — the *file name* of an export still comes from `GetDefaultStr(m_categoryData.m_name)`. A category whose name has no default-language entry is exported as the on-disk file name instead of its visible name. Harmless for OM's own files (with this PR the writer always emits a `<name>`), but it's the same rule this PR is removing everywhere else.

**KML round-trip now injects a default-language name.** For a bookmark with `name:ru` + `name:de` and no plain name, KML previously wrote `<name></name>` (expat calls no CharData, so nothing was parsed back). It now writes `<name>Eremitage</name>`, which `serdes.cpp:1274` parses into `m_name[kDefaultLang]` before ExtendedData adds ru/de — so export+import permanently pins the German name as the default one. Not a regression from this PR alone (it already happened whenever `<name>` was non-empty), but it now applies to a much wider set of bookmarks. If that matters, the parser could drop the `<name>`-derived default entry once `mwm:name` entries arrive.

**Alternative worth weighing: use the map language instead of a fixed priority list.** `GetNameForExport` is `GetPreferredBookmarkName` with a different language policy, and its last step, `GetLocalizedFeatureType`, is already device-locale dependent — so the export is not locale-independent anyway. Calling `GetPreferredBookmarkName(bmData, languages::GetCurrentMapLanguage())` from the writers would give one code path, no arbitrary lowest-code rule, and an exported name that always equals what the user sees in the app (`languages::GetCurrentMapLanguage()` lives in `platform/`, which `kml` already depends on via `type_utils.cpp`). Cons: two devices export the same collection under different names, and `GetPreferredBookmarkStr` returns `{}` for the ru+de case unless `m_name.size() == 1`, so `GetStringForExport` would still be needed as the tail. Current choice is defensible — flagging it because the mixed policy inside one function is easy to misread later.

Nits, take or leave:
- `GetStringForExport` drops the `int_name` comma truncation that `feature::GetBestName` applies ("junk names in Arabian island"), so a POI with `int_name="Foo, Bar"` and no default name now exports the full string. Probably the better behaviour for a file export, just noting the difference.
- The comment on line 1562 of `geojson_tests.cpp` explains why the *helper* parses JSON rather than what the escaping case checks — it belongs on `ExportedName` itself.
- Test data reviewed for language: `Mi lugar favorito`, `Скважина 12`, `Любимое место`, `Эрмитаж`, `Eremitage`, `Lieblingsort` are all correct and idiomatic; `Eremitage` is the right German exonym for the Hermitage.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes_common.cpp (hunk 7 above)
Comment (biodranik) [reply libs/kml/serdes_common.cpp]: Correcting my own snippet: `StringUtf8Multilang` already has a helper for exactly this — `IsAltOrOldName(int8_t)` (`libs/coding/string_utf8_multilang.hpp:88`, with `kAltNameCode = 53` / `kOldNameCode = 55`). Use it instead of the `GetLangIndex("alt_name")` calls I wrote, which do a string lookup per language per string:

```cpp
if (StringUtf8Multilang::IsAltOrOldName(lang))
  return 100 + lang;
```

`serdes_common.hpp` already includes that header, so no new dependency. Same as the fix proposed in the other review — the finding stands, only my suggested spelling of it was worse.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libs/kml/serdes.cpp (hunk 1 above)
Comment (biodranik) [reply libs/kml/serdes.cpp:434]: Confirmed on this branch — after a round trip all six strings come back with a `default` entry, five of them newly so.

Keeping it, because it turns out to be the point of the change rather than a side effect. `GetPreferredBookmarkStr` returns `{}` once `size() > 1` and neither the device language nor default/int_name/en is present, and unlike bookmarks, categories (`BookmarkCategory::GetName`, `bookmark.cpp:409`) and tracks (`SortTrackData::m_name`, `bookmark_manager.hpp:688`) have no further fallback — so such a name renders **empty** in the UI today. Measured with `{ru: Эрмитаж, de: Eremitage}`:

```
device    before       after
ru        Эрмитаж      Эрмитаж
de        Eremitage    Eremitage
fr        (empty)      Eremitage
default   (empty)      Eremitage
```

Now documented at `KmlParser::CharData` and pinned by the new `Kml_RoundTrip_PromotesNameWithoutDefaultLanguage`. Suppressing it would need per-string provenance for four elements plus per-compilation scoping, and would bring the blank names back.

File: libs/kml/serdes_common.cpp (hunk 2 above)
Comment (biodranik) [reply libs/kml/serdes_common.cpp]: Both applied: single pass with a `rank()` lambda, and `alt_name`/`old_name` pushed behind every real language (measurement in the reply on the other thread). The helpers also moved to `kml/type_utils`, next to the `GetPreferredBookmarkName` ladder they mirror.

File: libs/kml/serdes_common.hpp (hunk 3 above)
Comment (biodranik) [reply libs/kml/serdes_common.hpp]: Rewritten, and split into one comment per helper. The `GetNameForExport` one now states that a bookmark with no name at all falls back to `GetLocalizedFeatureType()`, so the same collection exports as "Castle" or "Замок" depending on the device UI language.

File: libs/kml/serdes_gpx.cpp (hunk 4 above)
Comment (biodranik) [reply libs/kml/serdes_gpx.cpp]: Right, rewritten. It now gives the rationale the copy actually serves: `m_customName` is what the place page, search and every exporter treat as the user's own name, so a human-authored GPX `<name>` belongs there, and a later rename replaces it instead of leaving two competing names.

File: libs/kml/kml_tests/geojson_tests.cpp (hunk 5 above)
Comment (biodranik) [reply libs/kml/kml_tests/geojson_tests.cpp]: Done — `TEST_EQUAL(ExportedName(typed), "historic-castle", ())`, with a comment noting that `platform::GetLocalizedTypeName()` is the identity in the test binary.

File: libs/kml/serdes_common.cpp (hunk 6 above)
Comment (biodranik) [reply libs/kml/serdes_common.cpp]: Applied. `GetStringForExport` returns `std::string_view`, `SaveStringWithCDATA` takes one, and only `GetNameForExport` materialises a `std::string` (it can return a temporary from `GetLocalizedFeatureType`). The two GeoJSON assignments use an explicit `std::string{...}`.

File: libs/kml/serdes_common.cpp (hunk 7 above)
Comment (biodranik) [reply libs/kml/serdes_common.cpp]: Confirmed and fixed. Probed on the branch before the fix:

```
{old_name, name:hi}  -> OLD NAME
{alt_name, name:kk}  -> ALT NAME
```

Both now export the real language. Reachable exactly as you describe — `Info::FormatNewBookmarkName()` copies every language of the feature. Using `StringUtf8Multilang::IsAltOrOldName()` per your follow-up. `ExportStringLanguageFallbacks` covers both directions, including the alt/old-only case, which still exports rather than dropping the string.

File: libs/kml/serdes_gpx.cpp (hunk 8 above)
Comment (biodranik) [reply libs/kml/serdes_gpx.cpp:298]: 1. Comment rewritten: it now says the KML parser copies the name for the same reason, but only for a plain single-language one — which is all this parser ever produces.

2. Checked, and the copy stays. Dropping it breaks exactly one test expectation and nothing else — every consumer resolves customName → name → type, and `Bookmark::GetCustomName()` has no callers at all. But it also buys nothing, because `bookmarks.kml` is our own storage and the KML rule at `serdes.cpp:1009` re-creates the copy on the next load:

```
customName after a KML round trip of a GPX-imported bookmark: size = 1, "Waypoint 1"
```

So the duplicate would come back within one app restart.

File: libs/kml/kml_tests/geojson_tests.cpp (hunk 9 above)
Comment (biodranik) [reply libs/kml/kml_tests/geojson_tests.cpp]: Removed. `if (error || exported.features.size() != 1)` became `TEST_EQUAL(exported.features.size(), 1, (json))`, and the ternary became `return name->second.get_string();`.

**Review by vng — APPROVED** [review]
